### ISSUE TITLE
adaptive_uniform_ratio is not a uniform floor: effective strength scales with num_envs and failure rate

### ISSUE THREAD

**Issue by linjiw** [mujocolab/mjlab#1153]
## Summary

In the tracking task's adaptive sampler, the uniform prior is combined with the failure statistic **additively**, and the failure statistic is an EMA of raw *counts*. The uniform term's share of the sampling distribution is therefore `ε / (Σq + ε)`, which shrinks as failures accumulate and as `num_envs` grows — rather than the fixed `ε` that `adaptive_uniform_ratio` implies.

Practical consequence: `adaptive_uniform_ratio` provides **no lower bound** on any bin's sampling probability, and the same configured value behaves very differently at different `num_envs`. Anyone who tuned it at one scale and then changed `num_envs` has silently changed their sampler.

## Where

`src/mjlab/tasks/tracking/mdp/commands.py`, `MotionCommand._adaptive_sampling`:

```python
sampling_probabilities = (
    self.bin_failed_count + self.cfg.adaptive_uniform_ratio / float(self.bin_count)
)
...
sampling_probabilities = sampling_probabilities / sampling_probabilities.sum()
```

and in `_update_command`:

```python
self.bin_failed_count = (
    self.cfg.adaptive_alpha * self._current_bin_failed
    + (1 - self.cfg.adaptive_alpha) * self.bin_failed_count
)
```

with `self._current_bin_failed[:] = torch.bincount(fail_bins, minlength=self.bin_count)`.

So `bin_failed_count` carries units of *episodes failed per update*, not a normalised distribution — and it scales with the number of environments.

## Derivation

Let `q` be the failure EMA over `N` bins and `ε = adaptive_uniform_ratio`:

```
p_i = (q_i + ε/N) / (Σq + ε)
```

The total mass contributed by the uniform term is

```
Σ_i (ε/N) / (Σq + ε)  =  ε / (Σq + ε)
```

which equals `ε` only when `Σq ≈ 1`. Since `q` is an EMA of counts, `Σq` tracks the mean number of failing episodes per update:

| Σq (mean failures/update) | uniform share at ε = 0.1 |
|---:|---:|
| 0.1 | 50% |
| 1 | 9.1% |
| 10 | 1.0% |
| 50 | 0.2% |

## Minimal reproduction

No simulator required — the sampler arithmetic alone shows it:

```python
import torch

N, eps = 100, 0.1
for scale in (0.1, 1.0, 10.0, 50.0):
    q = torch.zeros(N)
    q[0] = scale                      # failure mass concentrated on one bin
    additive = (q + eps / N) / (q + eps / N).sum()
    mixture  = (1 - eps) * (q / q.sum()) + eps / N
    print(f"Sum(q)={scale:5.1f}  additive: top1={additive.max():.3f}, "
          f"uniform share={eps/(q.sum()+eps):.3%}   "
          f"mixture: top1={mixture.max():.3f}, uniform share={eps:.1%}")
```

```
Sum(q)=  0.1  additive: top1=0.505, uniform share=50.000%   mixture: top1=0.901, uniform share=10.0%
Sum(q)=  1.0  additive: top1=0.910, uniform share=9.091%   mixture: top1=0.901, uniform share=10.0%
Sum(q)= 10.0  additive: top1=0.990, uniform share=0.990%   mixture: top1=0.901, uniform share=10.0%
Sum(q)= 50.0  additive: top1=0.998, uniform share=0.200%   mixture: top1=0.901, uniform share=10.0%
```

The additive form swings from 0.505 to 0.998 top-1 mass across a plausible range of failure rates; the mixture is invariant to it.

We hit this in a clip-level port of the same sampler (units = motion clips rather than time bins, 4096 envs, `ε = 0.1`, `adaptive_alpha = 0.01`): the distribution collapsed to a single unit holding 87–89% of all sampling mass in every seed, with normalised entropy falling to ~0.12–0.15, despite the nominal "10% uniform floor". Replacing the additive term with a convex mixture kept top-1 mass at ≤ 0.19.

## Suggested fix

Normalise before mixing, so `ε` is a scale-invariant coverage floor:

```python
q = self.bin_failed_count
q = q / q.sum() if float(q.sum()) > 0 else torch.full_like(q, 1.0 / self.bin_count)
sampling_probabilities = (1.0 - eps) * q + eps / self.bin_count
```

This changes behaviour for existing configs, so it may warrant a flag (e.g. `uniform_mixing: {"additive", "mixture"}`) defaulting to current behaviour for a release.

A documentation-only alternative: state that `adaptive_uniform_ratio` is an additive pseudo-count rather than a floor, and that its effective weight depends on `num_envs` and the failure rate, so it must be retuned whenever either changes.

Happy to open a PR for whichever option you prefer.

## Related

The same expression appears in BeyondMimic / `whole_body_tracking` at two sites; filed there as well for visibility — see cross-link below.

**Comment by linjiw** [comment]
Cross-link: the same expression appears in BeyondMimic / `whole_body_tracking` at `tasks/tracking/mdp/commands.py:217`, filed as HybridRobotics/whole_body_tracking#73 with the same derivation and repro.

**Comment by zjn20030811** [comment]
I’m preparing a focused fix for this issue: normalize the failure EMA into a probability distribution before applying `adaptive_uniform_ratio`, while retaining a compatibility option for the legacy additive behavior. I’ll add deterministic tests covering the zero-failure case, scale invariance, and the configured uniform floor, then run the relevant MotionCommand test suite.
